### PR TITLE
Refactorization of DB lock handle.

### PR DESCRIPTION
**Pull request by luisgf**
Refactorization of DB lock handle. open(),write(), read() function should be avoided to deal with files. The ANSI C functions fopen(), fwrite(), fread() should be used that is the "portable way"

### CODE REVIEW

**Comment by houz** [comment]
Is there a platform that requires these changes or is it just "because"? And even if there is, this most likely breaks the whole purpose of the code as it doesn't prevent using the database twice.

**Comment by luisgf** [comment]
No, the behaviour to limit database access to one instance of darktable is preserved, but the patch is a correction to be more portable. open() is a unix syscall that return a file descriptor, but fopen() is an ANSI C function that simply work with all systems.

**Comment by McBofh** [comment]
Both open() and fopen() are part of POSIX (see http://pubs.opengroup.org/onlinepubs/009695399/functions/open.html and http://pubs.opengroup.org/onlinepubs/009695399/functions/fopen.html), so any platform claiming POSIX adherence will have them.

I don't think your proposed change is necessary - do we really need to have buffered I/O for this lockfile?

**Comment by luisgf** [comment]
Yes @McBofh, both functions are part of POSIX but this limit the possibility of a program port to a Non-Posix system, like Windows.

File: src/common/database.c [hunk 1]
```diff
@@ -995,33 +996,25 @@ dt_database_t *dt_database_init(const char *alternative)
...
-            // the other process seems to no longer exist. unlink the .lock file and try again
-            unlink(db->lockfile);
+            // the other process seems to no longer exist. unlink the .lock file and try again            
             if(lock_tries < 5)
             {
-              close(fd);
+              int ret = fclose(fd);
+              if (ret) {
+                  fprintf(stderr, "[init] error opening the database lock file for reading\n");
+                  return NULL;
```
Comment (pedrocr) [inline src/common/database.c]: Is this really what you want? The lockfile hasn't failed to open, it failed to close.

File: src/common/database.c [hunk 2]
```diff
@@ -1030,20 +1023,21 @@ dt_database_t *dt_database_init(const char *alternative)
...
-      {
-        fprintf(stderr, "[init] error opening the database lock file for reading\n");
-      }
     }
+    }
+    else  // No lockfile exists - write our PID into new oneit
+    {
+        fd = fopen(db->lockfile, "w");
+        if (!fd) {
+            fprintf(stderr, "[init] error opening the database lock file for writing\n");
```
Comment (pedrocr) [inline src/common/database.c]: Should we also return NULL here, or is the other point wrong?

File: src/common/database.c [hunk 3]
```diff
@@ -1030,20 +1023,21 @@ dt_database_t *dt_database_init(const char *alternative)
...
+    {
+        fd = fopen(db->lockfile, "w");
+        if (!fd) {
+            fprintf(stderr, "[init] error opening the database lock file for writing\n");            
+        }
+        
+        if (fprintf(fd, "%d", getpid()))
+            db->lock_acquired = TRUE;  
+    }    
+    fclose(fd);
```
Comment (pedrocr) [inline src/common/database.c]: This can also fail, should we check this one as well?

File: src/common/database.c [hunk 4]
```diff
@@ -1030,20 +1023,21 @@ dt_database_t *dt_database_init(const char *alternative)
...
     }
+    }
+    else  // No lockfile exists - write our PID into new oneit
+    {
+        fd = fopen(db->lockfile, "w");
+        if (!fd) {
+            fprintf(stderr, "[init] error opening the database lock file for writing\n");            
+        }
+        
+        if (fprintf(fd, "%d", getpid()))
```
Comment (pedrocr) [inline src/common/database.c]: This is wrong, fprintf returns negative values (which evaluate to true) on error

File: src/common/database.c (hunk 2 above)
Comment (pedrocr) [inline src/common/database.c]: Yeah this is definitely wrong, if this triggers we would print something to stdout and then try the next line doing fprintf(NULL, ...)

**Comment by pedrocr** [comment]
As it stands this isn't really mergeable, the code is quite buggy and won't even work since the __WIN32__ ifdef is still there (so all this code would still be disabled under Windows anyway) and it still calls umask which is also POSIX.

**Comment by houz** [comment]
So, since this is not really needed for any supported platform and using fopen here is just wrong I am closing. I don't think that this could be made to work.

**Pull request closed without merging**